src/WinAppDriver/NewSessionHandler.cs

```diff
@@ -22,56 +24,86 @@ public object Handle(Dictionary<string, string> urlParams, string body, ref Sess
             NewSessionRequest request = JsonConvert.DeserializeObject<NewSessionRequest>(body);
             foreach (var kvp in request.DesiredCapabilities)
             {
-                Console.WriteLine("{0} = {1} ({2})", kvp.Key, kvp.Value, kvp.Value.GetType());
+                logger.Info("{0} = {1} ({2})", kvp.Key, kvp.Value, kvp.Value.GetType());
             }
 
             var caps = new Capabilities()
             {
                 PlatformName = (string)request.DesiredCapabilities["platformName"],
                 PackageName = (string)request.DesiredCapabilities["packageName"],
-                App = (string)request.DesiredCapabilities["app"],
+                App = request.DesiredCapabilities.ContainsKey("app") ? (string)request.DesiredCapabilities["app"] : null,
                 MD5 = request.DesiredCapabilities.ContainsKey("MD5") ? (string)request.DesiredCapabilities["MD5"] : null
             };
 
             IStoreApplication app = new StoreApplication(caps.PackageName, this.utils);
 
-            if (caps.MD5 != null && caps.MD5 == app.GetLocalMD5())
+            if (app.IsInstalled())

```

Comment:
The implementation here is too complicated. Most of the logic could be extracted to a separated collaborator - package installer.